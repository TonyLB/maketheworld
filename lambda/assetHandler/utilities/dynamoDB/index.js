const { GetItemCommand, PutItemCommand, QueryCommand, BatchWriteItemCommand } = require("@aws-sdk/client-dynamodb")
const { marshall, unmarshall } = require("@aws-sdk/util-dynamodb")
const { v4: uuidv4 } = require("uuid")

const { TABLE_PREFIX } = process.env;
const ephemeraTable = `${TABLE_PREFIX}_ephemera`
const permanentsTable = `${TABLE_PREFIX}_permanents`

const batchWriteDispatcher = (dbClient, { table, items }) => {
    const groupBatches = items
        .reduce((({ current, requestLists }, item) => {
            if (current.length > 19) {
                return {
                    requestLists: [ ...requestLists, current ],
                    current: [item]
                }
            }
            else {
                return {
                    requestLists,
                    current: [...current, item]
                }
            }
        }), { current: [], requestLists: []})
    const batchPromises = [...groupBatches.requestLists, groupBatches.current]
        .filter((itemList) => (itemList.length))
        .map((itemList) => (dbClient.send(new BatchWriteItemCommand({ RequestItems: {
            [table]: itemList
        } }))))
    return Promise.all(batchPromises)
}

//
// TODO:  Error handling to respond if the DynamoDB service throws an error
//
const replaceItem = async (dbClient, item) => {
    const putItem = new PutItemCommand({
        TableName: permanentsTable,
        Item: marshall(item, { removeUndefinedValues: true })
    })
    await dbClient.send(putItem)
}

const replaceRangeByDataCategory = async (dbClient, DataCategory, items, equalityFunction) => {
    const query = new QueryCommand({
        TableName: permanentsTable,
        KeyConditionExpression: "DataCategory = :dc",
        ExpressionAttributeValues: marshall({ ":dc": DataCategory }),
        IndexName: 'DataCategoryIndex'
    })
    const { Items: currentItems = [] } = await dbClient.send(query)
    const currentItemsByPermanentId = currentItems.map(unmarshall).reduce((previous, { PermanentId, ...rest }) => ({ ...previous, [PermanentId]: { PermanentId, ...rest }}), {})
    const scopedToPermanent = currentItems.map(unmarshall).filter(({ scopedId }) => (scopedId)).reduce((previous, { scopedId, PermanentId }) => ({ ...previous, [scopedId]: PermanentId }), {})
    //
    // Use a quick mutable map to track state as we combine several array
    // conditions
    //
    let stateByPermanentId = Object.values(currentItemsByPermanentId).reduce(
        (previous, { PermanentId }) => ({
            ...previous,
            [PermanentId]: {
                DeleteRequest: {
                    Key: marshall({ PermanentId, DataCategory })
                }
            }
        }), {})
    items.forEach((item) => {
        const { key, isGlobal, ...rest } = item
        const targetPermanentId = isGlobal ? `ROOM#${key}` : scopedToPermanent[key]
        if (targetPermanentId && equalityFunction(currentItemsByPermanentId?.[targetPermanentId] ?? {}, { PermanentId: targetPermanentId, scopedId: key, ...rest })) {
            stateByPermanentId[targetPermanentId] = 'ignore'
        }
        else {
            const newPermanentId = targetPermanentId || `ROOM#${uuidv4()}`
            stateByPermanentId[newPermanentId] = {
                PutRequest: {
                    Item: marshall({
                        DataCategory,
                        PermanentId: newPermanentId,
                        scopedId: key,
                        ...rest
                    }, { removeUndefinedValues: true })
                }
            }
        }
    })

    await batchWriteDispatcher(dbClient, {
        table: permanentsTable,
        items: Object.values(stateByPermanentId).filter((value) => (value !== 'ignore'))
    })
}

exports.replaceItem = replaceItem
exports.replaceRangeByDataCategory = replaceRangeByDataCategory
