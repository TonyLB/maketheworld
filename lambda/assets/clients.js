import { S3Client, GetObjectCommand as GetObjectCommandOriginal } from "@aws-sdk/client-s3"
import { EventBridgeClient } from "@aws-sdk/client-eventbridge"

export const ebClient = new EventBridgeClient({ region: process.env.AWS_REGION })

// export const s3Client = new S3Client(params)

export class GetObjectCommand extends GetObjectCommandOriginal {}
