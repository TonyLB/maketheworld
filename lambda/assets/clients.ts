import { GetObjectCommand as GetObjectCommandOriginal } from "@aws-sdk/client-s3"
import { EventBridgeClient, PutEventsCommand } from "@aws-sdk/client-eventbridge"
import { apiClient as apiClientImport } from "@tonylb/mtw-utilities/dist/apiManagement/apiManagementClient"
import { SFNClient } from "@aws-sdk/client-sfn"

export const ebClient = new EventBridgeClient({ region: process.env.AWS_REGION })
export const sfnClient = new SFNClient({ region: process.env.AWS_REGION })

// export const s3Client = new S3Client(params)

export class GetObjectCommand extends GetObjectCommandOriginal {}

export const apiClient = {
    send: (message: any) => {
        return apiClientImport.send(message)
            .catch(async (err: any) => {
                if (err.name === 'GoneException' || err.name === 'BadRequestException') {
                    await ebClient.send(new PutEventsCommand({
                        Entries: [{
                            EventBusName: process.env.EVENT_BUS_NAME,
                            Source: 'mtw.diagnostics',
                            DetailType: 'Force Disconnect',
                            Detail: JSON.stringify({
                                connectionId: message.ConnectionId
                            })
                        }]
                    }))
                }
                else {
                    console.log(`Error: ${err.name}`)
                    throw err
                }
            })
    }
}
