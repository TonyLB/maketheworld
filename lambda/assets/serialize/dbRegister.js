import { replaceItem, mergeIntoDataRange } from '../utilities/dynamoDB/index.js'

export const dbRegister = async (dbClient, { fileName, translateFile, importTree, scopeMap, assets }) => {
    const asset = assets.find(({ tag }) => (tag === 'Asset'))
    if (asset && asset.key) {
        await Promise.all([
            replaceItem(dbClient, {
                AssetId: `ASSET#${asset.key}`,
                DataCategory: 'Meta::Asset',
                fileName,
                translateFile,
                importTree,
                name: asset.name,
                description: asset.description,
                player: asset.player
            }),
            mergeIntoDataRange({
                dbClient,
                table: 'assets',
                search: { DataCategory: `ASSET#${asset.key}` },
                items: assets
                    .filter(({ tag }) => (tag === 'Room'))
                    .map(({ tag, ...rest }) => (rest)),
                mergeFunction: ({ current, incoming }) => {
                    if (!incoming) {
                        return 'delete'
                    }
                    if (!current) {
                        const { key, isGlobal, ...rest } = incoming
                        return {
                            scopedId: key,
                            ...rest
                        }
                    }
                    //
                    // TODO: When Room entries are expanded to store more than the sheer fact of their
                    // existence (likely as part of Map storage), extend this function to compensate
                    // by testing whether an update is needed
                    //
                    return 'ignore'
                },
                extractKey: (item) => {
                    if (item.isGlobal) {
                        return `ROOM#${item.key}`
                    }
                    if (scopeMap[item.key]) {
                        return `ROOM#${scopeMap[item.key]}`
                    }
                    console.log(`ERROR:  ScopeMap in dbRegister has no entry for ${item.key}`)
                    return `ROOM#${uuidv4()}`
                }
            })
        ])
    }
    const character = assets.find(({ tag }) => (tag === 'Character'))
    if (character && character.key) {
        await Promise.all([
            replaceItem(dbClient, {
                AssetId: `CHARACTER#${scopeMap[character.key]}`,
                DataCategory: 'Meta::Character',
                fileName,
                translateFile,
                scopedId: character.key,
                ...(['Name', 'Pronouns', 'FirstImpression', 'OneCoolThing', 'Outfit', 'player']
                    .reduce((previous, label) => ({ ...previous, [label]: character[label] }), {}))
            })
        ])
    }

}
