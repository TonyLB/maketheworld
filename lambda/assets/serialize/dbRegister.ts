import { v4 as uuidv4 } from 'uuid'
import { assetDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index.js'
import { AssetKey } from '@tonylb/mtw-utilities/dist/types.js'
import AssetWorkspace from '@tonylb/mtw-asset-workspace/dist/index.js'
import { isNormalAsset, isNormalComponent, isNormalExit, NormalForm, isNormalCharacter, NormalItem, isNormalMap, isNormalRoom, isNormalFeature, NormalReference, isNormalImport } from '@tonylb/mtw-wml/dist/normalize/baseClasses.js'
import { EphemeraCharacterId, EphemeraError } from '@tonylb/mtw-interfaces/dist/baseClasses.js'
import internalCache from '../internalCache'
import messageBus from '../messageBus'
import setEdges from '@tonylb/mtw-utilities/dist/graphStorage/update/setEdges'
import { graphDBHandler } from '../internalCache/graph'
import { graphStorageDB } from './graphCache'
import { CharacterKey } from '@tonylb/mtw-utilities/dist/types'

const tagRenderLink = (normalForm) => (renderItem) => {
    if (typeof renderItem === 'object') {
        if (renderItem.tag === 'Link') {
            return {
                ...renderItem,
                targetTag: normalForm[renderItem.to]?.tag
            }
        }
    }
    return renderItem
}

const denormalizeExits = (normalForm: NormalForm) => (contents: NormalReference[]) => {
    const exitTags = contents.filter(({ tag }) => (tag === 'Exit'))
    const exits = exitTags.map(({ key }) => (normalForm[key])).filter(isNormalExit)
    return exits.map(({ name, to }) => ({ name, to }))
}

const noConditionContext = ({ contextStack }) => (!contextStack.find(({ tag }) => (tag === 'If')))

const itemRegistry = (normalForm: NormalForm) => (item: NormalItem) => {
    if (isNormalMap(item)) {
        const mapDefaultAppearances = item.appearances
            .filter(noConditionContext)
            .map(({ rooms }) => ({
                rooms: rooms.map(({ key, ...rest }) => {
                    const lookup = normalForm[key]
                    if (!isNormalRoom(lookup)) {
                        throw new EphemeraError(`Invalid item in map room lookup: ${key}`)
                    }
                    return {
                        ...rest,
                        name: (lookup.appearances || [])
                            .filter(noConditionContext)
                            .map<string>(({ name }) => ((name || []).map((namePart) => (namePart.tag === 'String' ? namePart.value : '')).join('')))
                            .join('')
                    }
                })
            }))
        return {
            tag: item.tag,
            key: item.key,
            defaultAppearances: mapDefaultAppearances
        }
    }
    if (isNormalRoom(item)) {
        return {
            tag: item.tag,
            key: item.key,
            defaultAppearances: item.appearances
                .filter(noConditionContext)
                .filter(({ contents= [], render = [], name = [] }) => (contents.length > 0 || render.length > 0 || name.length > 0))
                .map(({ contents = [], render = [], name = [] }) => ({
                    exits: denormalizeExits(normalForm)(contents),
                    render: render.map(tagRenderLink(normalForm)),
                    name: name.map((item) => (item.tag === 'String' ? item.value : '')).join('')
                }))
        }
    }
    if (isNormalFeature(item)) {
        return {
            tag: item.tag,
            key: item.key,
            defaultAppearances: item.appearances
                .filter(noConditionContext)
                .filter(({ contents= [], render = [], name = [] }) => (contents.length > 0 || render.length > 0 || name.length > 0))
                .map(({ render, name = [] }) => ({
                    render: (render || []).map(tagRenderLink(normalForm)),
                    name: name.map((item) => (item.tag === 'String' ? item.value : '')).join('')
                }))
        }
    }
}

export const dbRegister = async (assetWorkspace: AssetWorkspace): Promise<void> => {
    const { address } = assetWorkspace
    // if (assetWorkspace.status.json !== 'Clean') {
    //     await assetWorkspace.loadJSON()
    // }
    const assets: NormalForm = assetWorkspace.normal || {}
    const asset = Object.values(assets).find(isNormalAsset)
    if (asset && asset.key) {
        const defaultExits = Object.values(assets)
            .filter(isNormalExit)
            .filter(({ appearances }) => ((appearances || []).find(noConditionContext)))
            .map(({ name, to, from }) => ({ name, to, from }))
        const defaultNames = Object.values(assets)
            .filter(isNormalComponent)
            .map(({ tag, key, appearances }) => ({
                tag,
                key,
                name: (appearances || [])
                    .filter(noConditionContext)
                    .map(({ name = [] }) => (name.map((item) => (item.tag === 'String' ? item.value : '')).join('')))
                    .join('')
            }))
            .filter(({ name }) => (Boolean(name)))
            .map(({ tag, key, name }) => ({ [key]: { tag, name } }))
            .reduce((previous, entry) => (Object.assign(previous, entry)), {})
        const updatedLibraryAssets = {
            [AssetKey(asset.key)]: {
                AssetId: AssetKey(asset.key),
                scopedId: asset.key,
                Story: asset.Story,
                instance: asset.instance,
            }
        }
        const updatedPlayerAssets = {
            [asset.key]: {
                AssetId: asset.key,
                scopedId: asset.key,
                Story: asset.Story,
                instance: asset.instance,
            }
        }
        const updateLibraryPromise = address.zone === 'Personal'
            ? internalCache.PlayerLibrary.set(address.player, {
                Assets: updatedPlayerAssets,
                Characters: {}
            })
            : address.zone === 'Library'
                ? internalCache.Library.set({
                    Assets: updatedLibraryAssets,
                    Characters: {}
                })
                : Promise.resolve({})
        await Promise.all([
            setEdges({ internalCache: internalCache._graphCache, dbHandler: graphStorageDB })(
                AssetKey(asset.key),
                Object.values(assets)
                    .filter(isNormalImport)
                    .map(({ from }) => ({ target: AssetKey(from), context: '' })),
                'back'
            ),
            assetDB.putItem({
                AssetId: AssetKey(asset.key),
                DataCategory: `Meta::Asset`,
                address,
                Story: asset.Story,
                instance: asset.instance,
                importTree: [],
                defaultExits,
                defaultNames,
                zone: address.zone,
                ...(address.zone === 'Personal' ? { player: address.player } : {})
            }),
            assetDB.merge({
                query: {
                    IndexName: 'DataCategoryIndex',
                    Key: {
                        DataCategory: AssetKey(asset.key)
                    }
                },
                items: asset.instance
                    ? []
                    : Object.values(assets)
                        .filter(({ tag }) => (['Room', 'Feature', 'Map'].includes(tag)))
                        .map(itemRegistry(assets)) as any,
                mergeFunction: ({ current, incoming }) => {
                    if (!incoming) {
                        return 'delete'
                    }
                    if (!current || (JSON.stringify(current) !== JSON.stringify(incoming))) {
                        const { tag, key, ...rest } = incoming
                        return {
                            scopedId: key,
                            ...rest
                        }
                    }

                    return 'ignore'
                },
                extractKey: ({ tag, key }) => {
                    let prefix = ''
                    switch(tag) {
                        case 'Feature':
                        case 'Map':
                        case 'Bookmark':
                        case 'Message':
                        case 'Moment':
                            prefix = tag.toUpperCase()
                            break
                        default:
                            prefix = 'ROOM'
                    }
                    if (assetWorkspace._isGlobal) {
                        return `${prefix}#${key}`
                    }
                    if (assetWorkspace.namespaceIdToDB[key]) {
                        return assetWorkspace.namespaceIdToDB[key]
                    }
                    console.log(`ERROR:  ScopeMap in dbRegister has no entry for ${key}`)
                    return `${prefix}#${uuidv4()}`
                }
            }),
            updateLibraryPromise
        ])
    }
    const character = Object.values(assets).find(isNormalCharacter)
    if (character && character.key) {
        const images = (character.images || [])
            .map((image) => (assetWorkspace.properties[image]?.fileName))
            .filter((image) => (image))
        const updatedCharacters = {
            [assetWorkspace.namespaceIdToDB[character.key]]: {
                CharacterId: assetWorkspace.namespaceIdToDB[character.key] as EphemeraCharacterId,
                Name: character.Name,
                FirstImpression: character.FirstImpression,
                OneCoolThing: character.OneCoolThing,
                Pronouns: character.Pronouns,
                Outfit: character.Outfit,
                fileName: '',
                fileURL: images[0] || undefined,
                scopedId: character.key,        
            }
        }
        const updateLibraryPromise = address.zone === 'Personal'
            ? internalCache.PlayerLibrary.set(address.player, {
                Assets: {},
                Characters: updatedCharacters
            })
            : address.zone === 'Library'
                ? internalCache.Library.set({
                    Assets: {},
                    Characters: updatedCharacters
                })
                : Promise.resolve({})
        await Promise.all([
            setEdges({ internalCache: internalCache._graphCache, dbHandler: graphStorageDB })(
                CharacterKey(character.key),
                Object.values(assets)
                    .filter(isNormalImport)
                    .map(({ from }) => ({ target: AssetKey(from), context: '' })),
                'back'
            ),
            assetDB.putItem({
                AssetId: assetWorkspace.namespaceIdToDB[character.key],
                DataCategory: `Meta::Character`,
                address,
                zone: address.zone,
                Name: character.Name,
                FirstImpression: character.FirstImpression,
                OneCoolThing: character.OneCoolThing,
                Pronouns: character.Pronouns,
                Outfit: character.Outfit,
                images,
                scopedId: character.key,
                ...(address.zone === 'Personal' ? { player: address.player } : {})
            }),
            updateLibraryPromise
        ])
        if (address.zone === 'Personal') {
            messageBus.send({
                type: 'PlayerInfo',
                player: address.player
            })

        }
        if (address.zone === 'Library') {
            messageBus.send({ type: 'LibraryUpdate' })
        }
    }

}
