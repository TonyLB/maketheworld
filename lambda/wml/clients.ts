import { EventBridgeClient } from "@aws-sdk/client-eventbridge"
import { S3Client } from "@aws-sdk/client-s3"
import { SFNClient } from "@aws-sdk/client-sfn"
import { SNSClient } from "@aws-sdk/client-sns"
import AWSXRay from 'aws-xray-sdk'

const params = { region: process.env.AWS_REGION }
export const ebClient = AWSXRay.captureAWSv3Client(new EventBridgeClient(params))
export const sfnClient = AWSXRay.captureAWSv3Client(new SFNClient(params))
export const snsClient = AWSXRay.captureAWSv3Client(new SNSClient(params))
export const s3Client = AWSXRay.captureAWSv3Client(new S3Client(params))
