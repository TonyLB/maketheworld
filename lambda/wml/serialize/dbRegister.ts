import { assetDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index.js'
import { AssetKey } from '@tonylb/mtw-utilities/dist/types.js'
import ReadOnlyAssetWorkspace from '@tonylb/mtw-asset-workspace/dist/readOnly'
import { isNormalAsset, NormalForm, isNormalCharacter, isNormalImport } from '@tonylb/mtw-wml/dist/normalize/baseClasses'
import { graphCache, graphStorageDB } from './graphCache'
import { CharacterKey } from '@tonylb/mtw-utilities/dist/types'
import GraphUpdate from '@tonylb/mtw-utilities/dist/graphStorage/update'
import { snsClient } from '../clients'
import { PublishCommand } from '@aws-sdk/client-sns'

const { FEEDBACK_TOPIC } = process.env

export const dbRegister = async (assetWorkspace: ReadOnlyAssetWorkspace): Promise<void> => {
    const { address } = assetWorkspace
    const assets: NormalForm = assetWorkspace.normal || {}
    const asset = Object.values(assets).find(isNormalAsset)
    if (asset && asset.key) {
        const graphUpdate = new GraphUpdate({ internalCache: graphCache, dbHandler: graphStorageDB })
        graphUpdate.setEdges([{
            itemId: AssetKey(asset.key),
            edges: Object.values(assets)
                .filter(isNormalImport)
                .map(({ from }) => ({ target: AssetKey(from), context: '' })),
            options: { direction: 'back' }
        }])
        await Promise.all([
            graphUpdate.flush(),
            assetDB.putItem({
                AssetId: AssetKey(asset.key),
                DataCategory: `Meta::Asset`,
                address,
                Story: asset.Story,
                instance: asset.instance,
                zone: address.zone,
                ...(address.zone === 'Personal' ? { player: address.player } : {})
            })
        ])
    }
    const character = Object.values(assets).find(isNormalCharacter)
    if (character && character.key) {
        const images = (character.images || [])
            .map((image) => (assetWorkspace.properties[image]?.fileName))
            .filter((image) => (image))
        const graphUpdate = new GraphUpdate({ internalCache: graphCache, dbHandler: graphStorageDB })
        graphUpdate.setEdges([{
            itemId: CharacterKey(character.key),
            edges: Object.values(assets)
                .filter(isNormalImport)
                .map(({ from }) => ({ target: AssetKey(from), context: '' })),
            options: { direction: 'back' }
        }])
        await Promise.all([
            graphUpdate.flush(),
            assetDB.putItem({
                AssetId: assetWorkspace.namespaceIdToDB[character.key],
                DataCategory: `Meta::Character`,
                address,
                zone: address.zone,
                Name: character.Name,
                FirstImpression: character.FirstImpression,
                OneCoolThing: character.OneCoolThing,
                Pronouns: character.Pronouns,
                Outfit: character.Outfit,
                images,
                scopedId: character.key,
                ...(address.zone === 'Personal' ? { player: address.player } : {})
            })
        ])
        if (address.zone === 'Personal') {
            await snsClient.send(new PublishCommand({
                TopicArn: FEEDBACK_TOPIC,
                Message: JSON.stringify({ player: address.player }),
                MessageAttributes: {
                    Type: { DataType: 'String', StringValue: 'PlayerInfo' }
                }
            }))
        }
        if (address.zone === 'Library') {
            await snsClient.send(new PublishCommand({
                TopicArn: FEEDBACK_TOPIC,
                Message: '{}',
                MessageAttributes: {
                    Type: { DataType: 'String', StringValue: 'LibraryUpdate' }
                }
            }))
        }
    }

}
