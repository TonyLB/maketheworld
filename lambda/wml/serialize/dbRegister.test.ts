import { jest, describe, it, expect } from '@jest/globals'

jest.mock('@tonylb/mtw-utilities/dist/dynamoDB/index')
import { assetDB } from '@tonylb/mtw-utilities/dist/dynamoDB/index'

jest.mock('@tonylb/mtw-utilities/dist/graphStorage/update', () => {
    return jest.fn().mockImplementation(() => {
        return {
            setEdges: mockSetEdges,
            flush: jest.fn()
        }
    })
})
import GraphUpdate from '@tonylb/mtw-utilities/dist/graphStorage/update'

jest.mock('../clients')
import { snsClient } from '../clients'

import { dbRegister } from './dbRegister'

const assetDBMock = assetDB as jest.Mocked<typeof assetDB>
const snsClientMock = snsClient as jest.Mocked<typeof snsClient>
const mockSetEdges = jest.fn()

describe('dbRegister', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.restoreAllMocks()
        mockSetEdges.mockClear()
    })
    it('should put a single element for a Character file', async () => {
        await dbRegister({
            address: {
                fileName: 'test',
                zone: 'Library',
            },
            namespaceIdToDB: { TESS: 'CHARACTER#12345' },
            status: {
                json: 'Clean'
            },
            properties: {
                TESSIcon: { fileName: 'IMAGE-123' }
            },
            normal: {
                TESS: {
                    tag: 'Character',
                    key: 'TESS',
                    Name: 'Tess',
                    fileURL: 'testIcon.png',
                    Pronouns: {
                        subject: 'she',
                        object: 'her',
                        possessive: 'her',
                        adjective: 'hers',
                        reflexive: 'herself'
                    },
                    FirstImpression: 'Frumpy Goth',
                    OneCoolThing: 'Fuchsia eyes',
                    Outfit: 'A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.',
                    images: ['TESSIcon']
                },
                TESSIcon: {
                    tag: 'Image',
                    key: 'TESSIcon'
                },
                'Import-0': {
                    tag: 'Import',
                    key: 'Import-0',
                    appearances: [{
                        contextStack: [{ key: 'TESS', tag: 'Character', index: 0 }],
                        contents: []
                    }],
                    from: 'primitives',
                }
            }
        } as any)
        expect(assetDBMock.putItem.mock.calls[0][0]).toMatchSnapshot()
        expect(mockSetEdges).toHaveBeenCalledTimes(1)
        expect(mockSetEdges).toHaveBeenCalledWith([{
            itemId: 'CHARACTER#TESS',
            edges: [{ target: 'ASSET#primitives', context: '' }],
            options: { direction: 'back' }
        }])
    })

    it('should save meta for Asset type', async () => {
        await dbRegister({
            address: {
                fileName: 'test',
                zone: 'Library'
            },
            status: {
                json: 'Clean'
            },
            namespaceIdToDB: {
                Welcome: 'ROOM#12345'
            },
            normal: {
                TEST: {
                    tag: 'Asset',
                    key: 'TEST',
                    name: 'Test',
                    appearances: [{
                        contextStack: [],
                        contents: [{
                            tag: 'Map',
                            key: 'Village',
                            index: 0
                        },
                        {
                            tag: 'Room',
                            key: 'Welcome',
                            index: 0
                        },
                        {
                            tag: 'Feature',
                            key: 'clockTower',
                            index: 0
                        },
                        {
                            tag: 'Variable',
                            key: 'power',
                            index: 0
                        },
                        {
                            tag: 'Action',
                            key: 'togglePower',
                            index: 0
                        }]
                    }]
                },
                Village: {
                    tag: 'Map',
                    key: 'Village',
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        rooms: [
                            { key: 'Welcome', x: 0, y: 100 }
                        ]
                    }]
                },
                Welcome: {
                    tag: 'Room',
                    key: 'Welcome',
                    appearances: [{
                        name: [{ tag: 'String', value: 'Welcome' }],
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                },
                clockTower: {
                    tag: 'Feature',
                    key: 'clockTower',
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                },
                power: {
                    tag: 'Variable',
                    key: 'power',
                    default: true,
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                },
                togglePower: {
                    tag: 'Action',
                    key: 'togglePower',
                    src: 'power = !power',
                    appearances: [{
                        contextStack: [{ tag: 'Asset', key: 'TEST', index: 0 }],
                        contents: []
                    }]
                }
            }
        } as any)
        expect(assetDBMock.putItem.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should save meta for Story type', async () => {
        await dbRegister({
            address: {
                fileName: 'test',
                zone: 'Library'
            },
            status: {
                json: 'Clean'
            },
            namespaceIdToDB: {
                Welcome: 'ROOM#12345'
            },
            normal: {
                TEST: {
                    tag: 'Asset',
                    Story: true,
                    key: 'TEST',
                    name: 'Test',
                    zone: 'Library',
                    appearances: [{
                        contextStack: [],
                        contents: [{
                            tag: 'Room',
                            key: 'Welcome',
                            index: 0
                        },
                        {
                            tag: 'Variable',
                            key: 'power',
                            index: 0
                        },
                        {
                            tag: 'Action',
                            key: 'togglePower',
                            index: 0
                        }]
                    }]
                },
                Welcome: {
                    tag: 'Room',
                    key: 'Welcome',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                },
                power: {
                    tag: 'Variable',
                    key: 'power',
                    default: true,
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                },
                togglePower: {
                    tag: 'Action',
                    key: 'togglePower',
                    src: 'power = !power',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                }
            }
        } as any)
        expect(assetDBMock.putItem.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should save meta for instanced Story type', async () => {
        await dbRegister({
            address: {
                fileName: 'test',
                zone: 'Library'
            },
            status: {
                json: 'Clean'
            },
            namespaceIdToDB: {
                Welcome: 'ROOM#12345'
            },
            normal: {
                TEST: {
                    tag: 'Asset',
                    Story: true,
                    instance: true,
                    key: 'TEST',
                    name: 'Test',
                    zone: 'Library',
                    appearances: [{
                        contextStack: [],
                        contents: [{
                            tag: 'Room',
                            key: 'Welcome',
                            index: 0
                        },
                        {
                            tag: 'Variable',
                            key: 'power',
                            index: 0
                        },
                        {
                            tag: 'Action',
                            key: 'togglePower',
                            index: 0
                        }]
                    }]
                },
                Welcome: {
                    tag: 'Room',
                    key: 'Welcome',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                },
                power: {
                    tag: 'Variable',
                    key: 'power',
                    default: true,
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                },
                togglePower: {
                    tag: 'Action',
                    key: 'togglePower',
                    src: 'power = !power',
                    appearances: [{
                        contextStack: [{ key: 'TEST', tag: 'Asset', index: 0 }],
                        contents: [],
                    }]
                }
            }
        } as any)
        expect(assetDBMock.putItem.mock.calls[0][0]).toMatchSnapshot()
    })

    it('should save asset graph edges when asset has imports', async () => {
        await dbRegister({
            address: {
                fileName: 'test',
                zone: 'Library'
            },
            status: {
                json: 'Clean'
            },
            namespaceIdToDB: {},
            normal: {
                test: {
                    tag: 'Asset',
                    key: 'test',
                    name: 'test',
                    zone: 'Library',
                    appearances: [{
                        contextStack: [],
                        contents: [{
                            tag: 'Import',
                            key: 'Import-0',
                            index: 0
                        },
                        {
                            tag: 'Room',
                            key: 'VORTEX',
                            index: 1
                        }]
                    }]
                },
                'Import-0': {
                    tag: 'Import',
                    key: 'Import-0',
                    appearances: [{
                        contextStack: [{ key: 'test', tag: 'Asset', index: 0 }],
                        contents: [{
                            key: 'VORTEX',
                            tag: 'Room',
                            index: 0
                        }]
                    }],
                    from: 'primitives',
                    mapping: { VORTEX: { key: 'VORTEX', type: 'Room' } }
                },
                VORTEX: {
                    tag: 'Room',
                    key: 'VORTEX',
                    appearances: [{
                        contextStack: [
                            { key: 'test', tag: 'Asset', index: 0 },
                            { key: 'Import-0', tag: 'Import', index: 0 }
                        ],
                        contents: [],
                        render: [],
                        name: []
                    },
                    {
                        contextStack: [{ key: 'test', tag: 'Asset', index: 0 }],
                        contents: [],
                        render: [],
                        name: []
                    }]
                }
            }
        } as any)
        expect(mockSetEdges).toHaveBeenCalledTimes(1)
        expect(mockSetEdges).toHaveBeenCalledWith([{
            itemId: 'ASSET#test',
            edges: [{ target: 'ASSET#primitives', context: '' }],
            options: { direction: 'back' }
        }])
    })

})