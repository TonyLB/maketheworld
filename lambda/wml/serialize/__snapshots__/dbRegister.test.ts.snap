// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dbRegister should put a single element for a Character file 1`] = `
{
  "AssetId": "CHARACTER#12345",
  "DataCategory": "Meta::Character",
  "FirstImpression": "Frumpy Goth",
  "Name": "Tess",
  "OneCoolThing": "Fuchsia eyes",
  "Outfit": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
  "Pronouns": {
    "adjective": "hers",
    "object": "her",
    "possessive": "her",
    "reflexive": "herself",
    "subject": "she",
  },
  "address": {
    "fileName": "test",
    "zone": "Library",
  },
  "images": [
    "IMAGE-123",
  ],
  "scopedId": "TESS",
  "zone": "Library",
}
`;

exports[`dbRegister should save meta for Asset type 1`] = `
{
  "AssetId": "ASSET#TEST",
  "DataCategory": "Meta::Asset",
  "Story": undefined,
  "address": {
    "fileName": "test",
    "zone": "Library",
  },
  "instance": undefined,
  "zone": "Library",
}
`;

exports[`dbRegister should save meta for Story type 1`] = `
{
  "AssetId": "ASSET#TEST",
  "DataCategory": "Meta::Asset",
  "Story": true,
  "address": {
    "fileName": "test",
    "zone": "Library",
  },
  "instance": undefined,
  "zone": "Library",
}
`;

exports[`dbRegister should save meta for instanced Story type 1`] = `
{
  "AssetId": "ASSET#TEST",
  "DataCategory": "Meta::Asset",
  "Story": true,
  "address": {
    "fileName": "test",
    "zone": "Library",
  },
  "instance": true,
  "zone": "Library",
}
`;
